### PR TITLE
DM-10342: Switch to C++14

### PR DESCRIPTION
**Pull request by pschella**
Implements RFC-332

### CODE REVIEW

File: coding/cpp_style_guide.rst [hunk 1]
```diff
@@ -84,10 +84,17 @@ These tasks are made much simpler if the code is easily readable and well-docume
...
+-------------------------
 
-The official policy on the use of C++11 features is at :ref:`Policy on use of C++11/14 language features <style-guide-cpp-cpp-11-14>`.
+The C++11 standard and the C++14 improvements to it bring a number of useful language features that make the resulting code more expressive, easier to read, and safer.
+
+We follow the official: International Standard ISO/IEC 14882:2014(E) – Programming Language C++, without any compiler specific extensions.
+
+.. seealso::
+
+   - :ref:`pipelines:source-install-redhat-legacy` from the `LSST Science Pipelines <https://pipelines.lsst.io>`__ documentation.
```
Comment (ktlim) [inline coding/cpp_style_guide.rst]: While this used to be sufficient, the docs for 14.0 that specify a minimum of gcc 4.8 are _not_ sufficient now.  We should make that clear.

File: coding/cpp_style_guide.rst [hunk 2]
```diff
@@ -84,10 +84,17 @@ These tasks are made much simpler if the code is easily readable and well-docume
...
 
-2-2. We are writing C++11/14 with some restrictions.
-----------------------------------------------------
+2-2. We are writing C++14
+-------------------------
 
-The official policy on the use of C++11 features is at :ref:`Policy on use of C++11/14 language features <style-guide-cpp-cpp-11-14>`.
+The C++11 standard and the C++14 improvements to it bring a number of useful language features that make the resulting code more expressive, easier to read, and safer.
+
+We follow the official: International Standard ISO/IEC 14882:2014(E) – Programming Language C++, without any compiler specific extensions.
```
Comment (ktlim) [inline coding/cpp_style_guide.rst]: I think we should say that the minimum compiler level is gcc 6.3.1 (and clang x.x.x?) and that if there are divergences between what the standard says should be supported and what the compilers support, the intersection set of portable features is what we allow.

File: coding/cpp_style_guide.rst [hunk 3]
```diff
@@ -84,10 +84,17 @@ These tasks are made much simpler if the code is easily readable and well-docume
...
+
+.. seealso::
+
+   - :ref:`pipelines:source-install-redhat-legacy` from the `LSST Science Pipelines <https://pipelines.lsst.io>`__ documentation.
+   - :doc:`/services/lsst-dev` provides :ref:`instructions for using devtoolset-6 <lsst-dev-tools>` to obtain a more modern GCC on LSST cluster machines.
```
Comment (ktlim) [inline coding/cpp_style_guide.rst]: Are there similar instructions for the minimum XCode level for macOS machines?

Comment (pschella) [reply coding/cpp_style_guide.rst]: You need at least Xcode 6 on OS X version 10.9.4 (e.g clang 3.4). But I don't think there are instructions.

File: coding/cpp_style_guide.rst (hunk 2 above)
Comment (pschella) [reply coding/cpp_style_guide.rst]: Done

**Pull request merged**